src/core/statelisteners/searchlistener.js

```diff
@@ -0,0 +1,176 @@
+import { SearchParams } from '../../ui';
+import QueryTriggers from '../models/querytriggers';
+import StorageKeys from '../storage/storagekeys';
+
+/**
+ * SearchListener is responsible for running a vertical or universal search when
+ * the QUERY storage key is updated. It also sets the defaultInitialSearch, if
+ * there is no QUERY already present.
+ */
+export default class SearchListener {
+  constructor (systemConfig, config) {
+    this.storage = systemConfig.storage;
+    this.componentManager = systemConfig.componentManager;
+    this.core = systemConfig.core;
+    this.config = {
+      searchCooldown: 300,
+      verticalKey: undefined,
+      allowEmptySearch: true,
+      defaultInitialSearch: undefined,
+      ...config
+    };
+
+    /**
+     * Middleware functions to be run before a search. Can be either async or sync.
+     */
+    this.middleware = [];
+  }
+
+  /**
+   * Sets the query on page load, and registers the search on QUERY update storage listener.
+   *
+   * @returns {SearchListener}
+   */
+  init () {
+    this._setQueryOnLoad();
+    this.storage.registerListener({
+      storageKey: StorageKeys.QUERY,
+      eventType: 'update',
+      callback: query => this._handleQueryUpdate(query)
+    });
+    return this;
+  }
+
+  /**
+   * Sets a search query and corresponding QUERY_TRIGGER on initialization. A prepopluated query
+   * in the URL takes precendence over a defaultInitialSearch.
+   */
+  _setQueryOnLoad () {
+    const prepopulatedQuery = this.storage.get(StorageKeys.QUERY);
+    if (prepopulatedQuery != null) {
+      if (this.storage.get(StorageKeys.QUERY_TRIGGER) !== QueryTriggers.SUGGEST) {
+        this.storage.setWithPersist(StorageKeys.QUERY_TRIGGER, QueryTriggers.QUERY_PARAMETER);
+      }
+    } else if (this.config.defaultInitialSearch != null) {
+      if (!this.config.defaultInitialSearch && !this.config.allowEmptySearch) {
```

Comment:
Some of these if statements are a little hard to follow. Is it possible to add some explanatory boolean values?